### ISSUE TITLE
Prop-Types creates PATENTS dependency, meaning I cannot upgrade.

### ISSUE THREAD

**Issue by thejohnbackes** [bvaughn/react-virtualized#771]
Hi,

Our npm package validation process rejects anything with a PATENTS file in the root. This means if there is a dependency on a PATENTS file, the entire installation is rejected.

We are using an old version of React (15.3) from before this validation was put in place, which is still compatible with new versions of React Virtualized, so I would love to upgrade.

However, currently Prop-Types is listed as a dependency, and upgrades are rejected. Would you be open to moving Prop-Types to "peerDependencies", similar to the situation with React?

I'm happy to post a pull request.

Thanks in advance!

John

**Comment by thejohnbackes** [comment]
or better yet, would you be open to moving them to an optional dependency + devDependency, to reflect the fact that some people choose not to use them....

**Comment by gaearon** [comment]
>We are using an old version of React (15.3) from before this validation was put in place, which is still compatible with new versions of React Virtualized, so I would love to upgrade.

Do you mean you intend to never update React from 15.3?

**Comment by thejohnbackes** [comment]
depends on whether the PATENTS issue is resolved in a way acceptable to
LinkedIn legal 😢😢😢

**Comment by spicyj** [comment]
React 15.3 has the same license as prop-types does. Am I misunderstanding something? Is the only relevant issue deps vs peerDeps?

**Comment by thejohnbackes** [comment]
currently we are unable to use any FB OSS that is BSD+PATENTS

**Comment by spicyj** [comment]
The last version of React that was not BSD+PATENTS was React 0.10.0. But you're using React 15.3 which is under BSD+PATENTS?

**Comment by bvaughn** [comment]
My understanding is that John is stuck on 15.3 because that's the version of React they were using _before_ this validation step was put into place and now they are unable to upgrade. But to be clear, the same PATENTS file applies to 15.3 as to 16 beta and prop-types so I think this is a bit confusing.

**Comment by thejohnbackes** [comment]
yes, that's correct. but up until a recent version of Virtualized, I was
able to upgrade without creating a direct PATENTS dependency, because React
is a Peer dependency. But when PropTypes was added as a dependency, now I
can no longer upgrade (since R-Virtualized 9.7.1). Yes I agree it's
confusing. We are doing our best to keep the community here going while our
own legal reviews happen...

**Comment by thejohnbackes** [comment]
to clarify: the current version of React we are using was grandfathered.

**Comment by bvaughn** [comment]
Right. I understand from a technical perspective why upgrading `react-virtualized` is blocked by the recent `prop-types` dependency.

What I _don't_ understand is- why there's concern about pulling in `prop-types` with its PATENTS file given that you're already using `react` and `react-dom` with identical PATENTS files. It seems like an artificial problem that has been created.

**Comment by gaearon** [comment]
Since the requirement itself is pretty odd, I can propose an odd workaround: create a Webpack alias (if you use webpack) from `prop-types` to a custom file that reexports `React.PropTypes`. This export still exists in 15.3.

**Comment by thejohnbackes** [comment]
I would agree, except, considering React-Virtualized in isolation, until
9.7.1 there was no direct dependency on anything with a PATENTS because
React is a peerDependency. So React does not get installed by npm by
default.

**Comment by gaearon** [comment]
Are you building a library? It’s not clear how you can write a React app without using React.

**Comment by thejohnbackes** [comment]
:-) agree the requirement is odd. it's a byzantine corporate workaround
thing, I guess, and probably unique to our situation.

not sure that's going to work because I can't get Virtualized onto our
internal npm repo if it has a dependency on PATENTS. it's not so much an
issue with webpack as it is with the validation step of our internal npm
repo... le sigh... 😔

To be clear, we are using React, but cannot add anything *new* with PATENTS
as of February of this year...

**Comment by bvaughn** [comment]
Sure. You could use something like `preact` + `preact-compat` with `react-virtualized` if the patents clause was a no-go for you. It's just confusing to me that you're actually using React (with the same patents file) already.

I was trying to think of a way to blacklist the `prop-types` dependency so you can continue to use the older version of React (and `React.PropTypes`) without your CI/whatever trying to install the new dependency but I'm not sure. Maybe you could use a Git alias to redirect `prop-types` installs to a dummy project that just re-exports `React.PropTypes` or something?

**Comment by thejohnbackes** [comment]
> 

Are you building a library? It’s not clear how you can write a React app without using React.
--

Loose Coupling! :-p

**Comment by thejohnbackes** [comment]
> Maybe you could use a Git alias to redirect prop-types installs to a dummy project that just re-exports React.PropTypes or something?

It might be possible to do this by creating a custom npm repo... Maybe :-(

**Comment by gaearon** [comment]
Maybe you could use Yarn. It has a hacky way to "redirect" npm packages as far as I know. You could do something like

```js
  "prop-types": "npm:my-prop-types"
```

which would be your fork that re-exports `React.PropTypes`. LOL

**Comment by thejohnbackes** [comment]
what a mess...

**Comment by bvaughn** [comment]
NPM supports a similar thing
```json
  "prop-types": "git://github.com/foo/FakePropTypes.git",
```
I've never tried it, but maybe you could just redirect installs this way?

**Comment by thejohnbackes** [comment]
> NPM supports a similar thing
> 
>   "prop-types": "git://github.com/foo/FakePropTypes.git",
> I've never tried it, but maybe you could just redirect installs this way?

So our process is:
1. Make external library request
2. package is fetched and validated
3. package dependencies are fetched and validated <-- this is the step that fails, based on React-PropTypes
4. Add package (now cached internally) to npm <-- this is where I would make the change you propose, but I still would have a problem in step 3...
5. Push changes
6. CI pulls from internal repo to build and deploy...

**Comment by gaearon** [comment]
Looks like your best bet is to fork `react-virtualized` and republish it without `prop-types` dependency.

**Comment by thejohnbackes** [comment]
> Looks like your best bet is to fork react-virtualized and republish it without prop-types dependency.

That... would work.

**Comment by thejohnbackes** [comment]
Thanks guys! I didn't really think of that approach!

**Comment by bvaughn** [comment]
Thought of suggesting that earlier but refrained because I hate to see forks for reasons like this. They make bug fixes and sharing improvements harder in the future. ☹️

**Comment by gaearon** [comment]
It does seem like they’d have to fork everything anyway because the ecosystem needs to move to `prop-types`. Also they wouldn’t get React bugfixes which to me seems like a bigger issue in long term.

**Comment by thejohnbackes** [comment]
> Also they wouldn’t get React bugfixes which to me seems like a bigger
issue in long term.

And no Fiber. We are very aware of this...

**Comment by thejohnbackes** [comment]
> Thought of suggesting that earlier but refrained because I hate to see forks for reasons like this. They make bug fixes and sharing improvements harder in the future. ☹️

Fortunately this should be the only change required, so we should be able to keep up to date with current version otherwise. As long as Prop-Types doesn't significantly diverge from React.PropTypes...

**Comment by thejohnbackes** [comment]
just updating this to say this issue is no longer relevant! :+1: :+1: :1st_place_medal:

**Comment by bvaughn** [comment]
Huzzah! 🎉

**Comment by gaearon** [comment]
💃👯🕺🏿
